File: jenfengchang/309.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0) {
            return 0;
        }
        int cash_pre = 0, cash = 0, hold = -prices[0];
        for(const auto& price:prices) {
            int tmp = cash;
            cash = max(cash, hold + price);
            hold = max(hold,cash_pre - price);
            cash_pre = tmp;
        }
        return cash;
    }
};

File: zemiao/309.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        hold = -32768
        release = [-32768, -32768]
        for i in range(len(prices)):
            price = prices[i]
            newRelease = max(release[-1], hold+price, 0)
            hold = max(hold, release[-2]-price if i-2 >= 0 else -price)
            release = [release[-1], newRelease]
            
        return max(release + [0])